compiler/rustc_mir/src/transform/check_consts/validation.rs

```diff
@@ -860,6 +860,16 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
 
                 if is_lang_panic_fn(tcx, callee) {
                     self.check_op(ops::Panic);
+
+                    // an internal `panic()` call must have at least one argument
+                    if args[0].ty(&self.ccx.body.local_decls, tcx).kind() != TyKind::Str {
```

Comment:
I think this is not going to work, because the type is `&str`, and this check is for `str`. I think what you can do is to first match with a `ty::Ref(_, ty, _)` pattern to get the thing behind the reference, and then there's actually a convenience function `is_str()` that will do the `.kind() == ty::Str` for you.

Reply:
@oli-obk I believe this recurses through the projection list to find the root type; it's what @RalfJung recommended. I did see `.is_str()` but I think it may be in the wrong module because my IDE didn't suggest it; should I just elevate its visibility?

Reply:
This call goes through the `Place::projection_items()` to get the final type, I believe.

Reply:
yes, but the "final type" is a `&str`. The reason it goes through `Place::Projection_items()` is so that `panic!(*x.foo)` will give us whatever value is at the destination of `foo`, so e.g. if `x: SomeStruct` where `struct SomeStruct { foo: &&str }`.

Reply:
You can insert `trace!` statements to check the type and/or make your new `ops` struct have a `Ty<'tcx>` field allowing the error message to emit the type that was encountered. This may make debugging easier in case a test isn't behaving as you expect it to.

Reply:
> I think this is not going to work, because the type is &str, and this check is for str. I think what you can do is to first match with a ty::Ref(_, ty, _) pattern to get the thing behind the reference, and then there's actually a convenience function is_str() that will do the .kind() == ty::Str for you.

Oh, good catch. Currently the code requires the argument type to be `str` but we want it to be `&str`.